lib/charms/operator_libs_linux/v0/dnf.py

```diff
@@ -0,0 +1,239 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Abstractions for system's DNF package information and repositories."""
+
+import os
+import re
+import shutil
+import subprocess
+from enum import Enum
+from io import StringIO
+from typing import Dict, Optional, Union
+
+
+class ExecutionError(Exception):
+    """Raise when dnf encounters an execution error."""
+
+
+class _PackageState(Enum):
+    INSTALLED = "installed"
+    AVAILABLE = "available"
+    ABSENT = "absent"
+    UNKNOWN = "unknown"
+
+
+class PackageInfo:
+    """Dataclass representing DNF package information."""
+
+    def __init__(self, data: Dict[str, Union[str, _PackageState]]) -> None:
+        self._data = data
+
+    @property
+    def installed(self) -> bool:
+        """Determine if package is marked 'installed'."""
+        return self._data["state"] == _PackageState.INSTALLED
+
+    @property
+    def available(self) -> bool:
+        """Determine if package is marked 'available'."""
+        return self._data["state"] == _PackageState.AVAILABLE
+
+    @property
+    def absent(self) -> bool:
+        """Determine if package is marked 'absent'."""
+        return self._data["state"] == _PackageState.ABSENT
+
+    @property
+    def unknown(self) -> bool:
+        """Determine if package is marked 'unknown'."""
+        return self._data["state"] == _PackageState.UNKNOWN
+
+    @property
+    def name(self) -> str:
+        """Get name of package."""
+        return self._data["name"]
+
+    @property
+    def arch(self) -> Optional[str]:
+        """Get architecture of package."""
+        return self._data.get("arch", None)
+
+    @property
+    def epoch(self) -> Optional[str]:
+        """Get epoch of package."""
+        return self._data.get("epoch", None)
+
+    @property
+    def version(self) -> Optional[str]:
+        """Get version of package."""
+        return self._data.get("version", None)
+
+    @property
+    def full_version(self) -> Optional[str]:
+        """Get full version of package."""
+        if self.absent:
+            return None
+
+        full_version = [self.version, f"-{self.release}"]
+        if self.epoch:
+            full_version.insert(0, f"{self.epoch}:")
+
+        return "".join(full_version)
+
+    @property
+    def release(self) -> Optional[str]:
+        """Get release of package."""
+        return self._data.get("release", None)
+
+    @property
+    def repo(self) -> Optional[str]:
+        """Get repository package is from."""
+        return self._data.get("repo", None)
+
+
+def version() -> str:
+    """Get version of `dnf` executable."""
+    return StringIO(_dnf("--version")).readline().strip("\n")
+
+
+def installed() -> bool:
+    """Determine if the `dnf` executable is available on PATH."""
+    return shutil.which("dnf") is not None
+
+
+def update() -> None:
+    """Update all packages on the system."""
+    _dnf("update")
```

Comment:
Yeah this was something I caught over the weekend as well :sweat_smile:. Here is how I changed it:

```python3
def upgrade(*packages: Optional[str]) -> None:
    """Upgrade one or more packages.

    Args:
        *packages (Optional[str]):
            Packages to upgrade on system. If None, upgrade all packages.
    """
    if len(packages) == 0:
        _dnf("upgrade")
    else:
        _dnf("upgrade", *packages)
```

_Ignore the fact that I am using `if len(packages) == 0`; I saw your comment later down in the thread._

Regarding `dnf upgrade ...` vs `dnf update ...`, having two separate functions for both `upgrade` and `update` is intentional due to something I am cooking up within the HPC team. I am trying to have parity between a custom APT charm library and this dnf charm library. Basically I am going for dnf and apt to have the same method signatures with little as possible difference between them so that it is easier to abstract them within an mixin/parent class. You can see my "prototype" (I am still working on the specification for supporting _alternative_ charm bases) [here](https://github.com/NucciTheBoss/cleantest/blob/0.4.0-dev/src/cleantest/utils/apt.py).

Reply:
I'm not at all sold on trying to make them conform to a standard interface (unless we were to do that formally with a typing.Protocol). But even so, I don't think it's a wise idea *here*. Package managers are all slightly different and use different terminology, and I think we'd be providing people with a "false sense of commonality" to kind of half give them parity. I think in `dnf.py` we should make the names/signatures DNF-like, and in `apt.py` we should make them APT-like.

If on top of that we need an abstracted-away interface, we can design that separately in a separate PR and/or charm lib. But even things like package names differ between package managers / repos, so I don't actually see that being very doable.

Reply:
Yeh, agree. The user facing expectations of dnf and apt are really quite different, and these libraries should feel natural to people who know those tools - more a lightweight Python wrapper around each

Reply:
Dropped the update method in https://github.com/canonical/operator-libs-linux/pull/75/commits/d19923ec4e869150f7e257a6c55d8fcffffdaaf5